### PR TITLE
AMP Tracker class, etc

### PR DESCRIPTION
**Pull request by ofloveandhate**
I propose pulling in my AMP Tracker class, and all other consequential changes contained in my release branch.

### CODE REVIEW

**Comment by ofloveandhate** [comment]
I recognize in retrospect that many many of these commits should have been squashed.  Future commits will be much smaller in terms of number of commits.  For review, I recommend looking at the end product, not the intermediate.  In particular, I would like the focus to be on tracking.  

There is some content related to logging of data, which is in this PR as a placeholder for more sophisticated logging abilities in the future.  There are other capabilities in this PR as well, but again, the focus should be on tracking.

That said, there should obviously be review on all submitted code.

**Comment by nilayvaish** [comment]
Daniel, the pull request says that you want to merge 492 commits.  But it seems that a lot of these commits have already been merged into the develop branch.  For example, the very first commit in this pull request is: 

---------------------------------------------------------------------------------
commit d2d62f4fe894b2ef61890b38b92996ae5d8e8678
added the leaky *.sublime-workspace file to .gitignore. 

---------------------------------------------------------------------------------
But this has already been committed to the develop branch:

---------------------------------------------------------------------------------
commit 472278c88c18fd29043a234692fa5188de8659af
Author: daniel brake <danielthebrake@gmail.com>
Date:   Wed Aug 12 16:14:56 2015 -0400
added the leaky *.sublime-workspace file to .gitignore.

---------------------------------------------------------------------------------

Is it possible for you to edit the pull request so that only the commits that are not part of the current develop branch are part of the request?

Just a disclaimer.  I am using GitHub for the very first time, so I might have made some mistake in understanding how the pull requests work.

Thanks
Nilay

**Comment by ofloveandhate** [comment]
This is indeed a problem.  I thought I was in the clear.

The problem is that the duplicates for some reason have distinct commit
numbers.  While you and I recognize that these are the same commit, Git
thinks they are different for some reason.  I've tried repeatedly to
identify and rectify my mistake, and have consulted a number of people on
the topic, but continue to fail.

Can anyone help me?

**Comment by jbcolli2** [comment]
So I don't really have anything to help, but it seems like we need to figure out why the same commit has two different numbers.  I'm wondering if this is the kind of problem we can't fix right now, but we can change the way we do things to make sure we don't have huge pull requests in the future.

**Comment by nilayvaish** [comment]
I am going to test things with GitHub (most likely by creating some dummy
repository) and see how to avoid the situation we are in right now.  Wait
for a couple of days.

Nilay

**Comment by ofloveandhate** [comment]
tl;dr: I want to start over with the repo.  Not the code, just the repo.

--

I have good news and bad news.

The good news is that I've identified the source of the `duplicates'.  Jeb, remember the .DS_Store problem?  Well, to fix the problem I employed the [BFG Repo Cleaner](https://rtyley.github.io/bfg-repo-cleaner/), which lets one remove files from a repo entirely.  I removed the .DS_Store from every commit of the bertiniteam/b2 repo using this tool. 

The bad new is that history got rewritten, and the hashes of commits got changed.  So now it appears that the commits in bertiniteam/b2 are distinct from those in our own forks.  For each of us.  

I cannot undo this action, to the best of my knowledge.  While I acted in good faith removing the files using BFG, I damaged our collective history.  This wouldn't be a problem, except that we had active work going on.  

Alan and I are trying to figure out the way out of this problem.  Maybe someone else out there, knowing now the source of the problem, can help find a solution.  My gut feeling is to nuke the repo from orbit.  That is, I think I'd like to delete the current history of commits, re-structure the repo so that it looks like what we want it to look like, and then each of us re-forks (or whatever we need to do), brings in our current work, and we (I) never let this happen again.

Long story short, history got re-written -- I didn't understand how BFG, rebasing, and some other patterns in Git really work.  This causes the commits to appear as duplicates.  I think the solution to this problem is to erase the commit history for bertiniteam/b2 (including the super messy start, where a bunch of xcode temp files got committed), for us to save our individual current progress, and to start over.  

Sorry I caused this problem.  My bad, it definitely won't happen again.

I would love feedback on how to move forward.

**Comment by jbcolli2** [comment]
Dan, I do remember the .DS_Store problem, and I cleaned my repo fork with
the same thing which problem just complicated the problem.  I think
restarting the repo is a great idea.  This repo was started way back when
we were just sitting around the basement of Weber trying to figure out what
we were doing.  You're right in that it has tons of junk at the beginning.
I think since we know so much more of what we're doing now, we can start
over from scratch and have a clean repo for the real Bertini 2.0 code.

  I think the main trick is making sure nothing gets lost from our personal
forked repos when we refork, but that shouldn't be too much of a problem.

On Tue, Dec 15, 2015 at 9:07 AM, daniel brake <notifications@github.com>
wrote:

> tl;dr: I want to start over with the repo. Not the code, just the repo.
>
> I have good news and bad news.
>
> The good news is that I've identified the source of the `duplicates'. Jeb,
> remember the .DS_Store problem? Well, to fix the problem I employed the BFG
> Repo Cleaner <https://rtyley.github.io/bfg-repo-cleaner/>, which lets one
> remove files from a repo entirely. I removed the .DS_Store from every
> commit of the bertiniteam/b2 repo using this tool.
>
> The bad new is that history got rewritten, and the hashes of commits got
> changed. So now it appears that the commits in bertiniteam/b2 are distinct
> from those in our own forks. For each of us.
>
> I cannot undo this action, to the best of my knowledge. While I acted in
> good faith removing the files using BFG, I damaged our collective history.
> This wouldn't be a problem, except that we had active work going on.
>
> Alan and I are trying to figure out the way out of this problem. Maybe
> someone else out there, knowing now the source of the problem, can help
> find a solution. My gut feeling is to nuke the repo from orbit. That is, I
> think I'd like to delete the current history of commits, re-structure the
> repo so that it looks like what we want it to look like, and then each of
> us re-forks (or whatever we need to do), brings in our current work, and we
> (I) never let this happen again.
>
> Long story short, history got re-written -- I didn't understand how BFG,
> rebasing, and some other patterns in Git really work. This causes the
> commits to appear as duplicates. I think the solution to this problem is to
> erase the commit history for bertiniteam/b2 (including the super messy
> start, where a bunch of xcode temp files got committed), for us to save our
> individual current progress, and to start over.
>
> Sorry I caused this problem. My bad, it definitely won't happen again.
>
> I would love feedback on how to move forward.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/bertiniteam/b2/pull/17#issuecomment-164791889>.
>



-- 

J.B. Collins, Ph.D
West Texas A&M University
Department of Mathematics
CC 421D
806-651-5260

**Comment by ofloveandhate** [comment]
By common consensus, we have decided to do a reset on the Git repo for Bertini2.  The beginning commits are dirty, and we have experienced a divergence of commits due to use of BFG repo cleaner.

Please save your current work and get ready for a re-structure.  Soon you will be requested to delete your b2 fork, and re-fork after the re-structure is completed.  

A new pull request will be opened after the cleanup is complete.
